### PR TITLE
Add new field DisableBGPExport to IPPools API

### PR DESCRIPTION
**Pull request by lmm**
## Description

This PR adds the new field `DisableBGPExport` to the operator IPPool API.
This field allows you to configure the default IP pools so that they do not export their routes over BGP.

Related: https://github.com/projectcalico/calico/pull/6391

## For PR author

- [x] Tests for change.
- [x] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [x] Milestone set according to targeted release.
- [x] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

**Review by tmjd — APPROVED** [review]
I'm assuming the ippool.go update was just a copy from libcalico.
LGTM

File: pkg/apis/crd.projectcalico.org/v1/ippool.go [hunk 1]
```diff
@@ -54,36 +65,74 @@ type IPPoolSpec struct {
...
+	IPPoolAllowedUseTunnel                    = "Tunnel"
+)
+
 type VXLANMode string
 
 const (
 	VXLANModeNever       VXLANMode = "Never"
-	VXLANModeAlways      VXLANMode = "Always"
-	VXLANModeCrossSubnet VXLANMode = "CrossSubnet"
+	VXLANModeAlways                = "Always"
```
[the remark this answers is not in the document]
Comment (lmm) [reply pkg/apis/crd.projectcalico.org/v1/ippool.go]: > I'm assuming the ippool.go update was just a copy from libcalico.

Yes I copied the update from here: https://github.com/projectcalico/calico/blob/3f9a10136e5918ba70b9fdaf61e781c4a7789dac/api/pkg/apis/projectcalico/v3/ippool.go#L100

> I'm surprised CI isn't complaining about these not all having the type. I had to add it originally so static checks would pass.

Good eye 🔎 so we don't actually run the `static-checks` target anymore which is why this wasn't flagged. Running `make static-checks` myself I saw the issues you thought should be there and fixed them.

File: pkg/apis/crd.projectcalico.org/v1/ippool.go [hunk 2]
```diff
@@ -54,13 +65,35 @@ type IPPoolSpec struct {
...
+
+	// The block size to use for IP address assignments from this pool. Defaults to 26 for IPv4 and 122 for IPv6.
 	BlockSize int `json:"blockSize,omitempty"`
 
 	// Allows IPPool to allocate for a specific node by label selector.
 	NodeSelector string `json:"nodeSelector,omitempty" validate:"omitempty,selector"`
+
+	// Deprecated: this field is only used for APIv1 backwards compatibility.
+	// Setting this field is not allowed, this field is for internal use only.
+	IPIP *IPIPConfiguration `json:"ipip,omitempty" validate:"omitempty,mustBeNil"`
```
Comment (caseydavenport) [inline pkg/apis/crd.projectcalico.org/v1/ippool.go:79]: IMO we shouldn't be including these fields (IPIP / NatOutgoingV1 / AllowedUses) - they're not used by the operator code at all, and are deprecated internal fields from a past version of the API.

File: pkg/apis/crd.projectcalico.org/v1/ippool.go [hunk 3]
```diff
@@ -77,13 +110,29 @@ const (
...
 
-// IPPoolList contains a list of IPPool resources.
-type IPPoolList struct {
-	metav1.TypeMeta `json:",inline"`
-	metav1.ListMeta `json:"metadata"`
-	Items           []IPPool `json:"items"`
+const (
+	Undefined   EncapMode = ""
+	Always      EncapMode = "always"
+	CrossSubnet EncapMode = "cross-subnet"
```
Comment (caseydavenport) [inline pkg/apis/crd.projectcalico.org/v1/ippool.go:120]: These don't seem to be used either?

Comment (caseydavenport) [reply pkg/apis/crd.projectcalico.org/v1/ippool.go:120]: (Just on IPIPConfiguration, which is an outdated struct)

File: pkg/apis/crd.projectcalico.org/v1/ippool.go (hunk 2 above)
Comment (lmm) [reply pkg/apis/crd.projectcalico.org/v1/ippool.go]: Good point, I've rolled back the unnecessary changes and only left the `disableBGPExport` field on IPPoolSpec that we need: d28a2184260c1475ae0fa66f766d082a5a4dc2ba

File: pkg/apis/crd.projectcalico.org/v1/ippool.go (hunk 3 above)
Comment (lmm) [reply pkg/apis/crd.projectcalico.org/v1/ippool.go]: I removed all the unneeded changes.

File: api/v1/installation_types.go [hunk 4]
```diff
@@ -489,6 +489,11 @@ type IPPool struct {
 	// Default: 26 (IPv4), 122 (IPv6)
 	// +optional
 	BlockSize *int32 `json:"blockSize,omitempty"`
+
+	// DisableBGPExport specifies whether routes from this IP pool's CIDR are exported over BGP
+	// Default: false
```
Comment (caseydavenport) [inline api/v1/installation_types.go:494]: Are we missing defaulting for this field? The comment claims this defaults to false but I don't see the code for that. Could probably work with a kubebuilder default tag.

File: pkg/controller/migration/convert/ippools_test.go [hunk 5]
```diff
@@ -246,6 +246,9 @@ var _ = Describe("Convert network tests", func() {
 			Entry("v4 and v6 pool but no assigning v4", `"assign_ipv4": "false", "assign_ipv6": "true"`, "1.168.4.0/24", "ff00:0001::/24"),
 			Entry("v4 and v6 pool but no assigning v6", `"assign_ipv4": "true", "assign_ipv6": "false"`, "1.168.4.0/24", "ff00:0001::/24"),
 		)
+
+		trueValue := true
+		//falseValue := false
```
[the remark this answers is not in the document]
Comment (lmm) [reply pkg/controller/migration/convert/ippools_test.go]: Fixed

File: api/v1/installation_types.go (hunk 4 above)
Comment (lmm) [reply api/v1/installation_types.go:494]: I've added kubebuilder tag to do the defaulting as suggested 👍

**Review by caseydavenport — APPROVED** [review]

**Pull request merged**